### ISSUE TITLE
Pirania data usage metering

### ISSUE THREAD

**Issue by tbdinesh** [libremesh/lime-app#337]
We are excited that a new release of Pirania is around the corner and looking forward to it to test it in our community (COWmesh).

Ours is a diverse community of 2000 where we wish to bring local knowledge and community resource maps available to the community via captive portal. We have used Pirania for a while but now we have reverted to passwords at some nodes and some nodes being open to all. Our wish is to incentivise local learning using vouchers for internet - based on some subjective criteria we work out with the community.

0. What we are looking forward is for a community woman member (with a RaspberryPi or using phone) but who is not tech savvy (not command line or editing csv files) to help with voucher generation. Say a validity of a few hours or a few days per voucher.

There has been a high use of gaming on the the phones (mostly by boys in the area and in general https://duckduckgo.com/?q=pubg+india+boys) which requires us to be discreet on generating vouchers. While we want to encourage online learning (video conference and youtube lessons are examples) we notice that some boys are addicted to these games and sit around the nodes for hours at a time. The elders in the community are not happy to see this.

1. We wish for the women (and us) to get an insight to the amount of data used per voucher which can be a point for negotiating for the next voucher.  

2. Also it would be great to be able to generate vouchers that are valid only for certain hours of the day (say 8pm to 11pm only)

**Comment by mikerjensen** [comment]
Agree, and vouchers based on data traffic rather than time will allow networks which use mobile data bundles for their backhaul/upstream connection to more feasibly recover their costs. Capability to offer vouchers with time-based validity can be used for smoothing bandwidth demand by encouraging use in off-peak periods.  Perhaps a more difficult but valuable feature to implement would be the capability to issue different types of vouchers according to the speed/bandwidth needed by the user.

**Comment by tbdinesh** [comment]
For us, for now, we see it as a healthy start of a conversation between the women who will be helping with generating the vouchers ("responsible" community members) and the mischievous (generally boys who one-up in terms of gaming the system) without having to arbitrarily cut off their internet access or to have a blanket block them all as the alternative.

We are not doing mobile data bundles. We have 300mbps fiber at inbound (powerbeam) but it throttles down to avg of 20mbps at the end points. The speed for all drops considerably as the gamers and others who use OTT platforms (similar to Netflix) can considerably slowdown the speed for WfH and student needs.

**Comment by patogit** [comment]
How do we add the ability to make a greater variety of vouchers, without overwhelming users with basic skills?

Maybe on the voucher creation page, there could be an `Advanced` button to change to an interface with more options and some conditionally-visible fields:
- restrict speed: `yes` (or `no`)
  - maximum speed: `0.5` Mbps
- restrict time of day: `yes` (or `no`)
  - period 1
    - open: `06:00`
    - close: `07:00`
  - period 2
    - open: `18:30`
    - close: `20:00`
  - add more periods if desired
- restrict days of the week: `yes` (or `no`)
  - allow only on the following days: `checkbox for each day of the week` (maybe there are cultural considerations, for cultures that measure weeks with more or less than 7 days?) 
- limit time that voucher is valid after activation:
  - time amount: `1`
  - time unit: `hours`
- limit total traffic (i.e. total data download)
  - voucher download limit: `500`
  - voucher download limit unit: `MB`
- category of voucher:
  - choose from the list, or make a new category (for example: students, adults, community leadership, visitors)

There's also the matter of: What role can easy-to-implement technical restrictions play in preventing game addiction, and to what extent is human vigilance and company necessary?

(EDIT: voucher parameters edited a little on 2023-dec-11.)

**Comment by patogit** [comment]
@tbdinesh , How is your community doing these days? How are you all controlling access to your community network? None of these ideas have been implemented to make vouchers with more detailed parameters. For our community here in Ecuador, more detailed vouchers would be very useful.

I imagine that the process to implement this would require:
- making the interface to edit the vouchers, like what I described in my previous post.
- making Pirania internal code capable of enforcing the new rules. This means tracking time of day, day of week, bandwidth control, and data usage. Of those, the first three are parameters to set, and the fourth requires a counter for each voucher.

I do not know enough about LibreMesh programming to know exactly how to implement this.

Also, for situations where we want even more control over the use of the network, there could be an option to create an allow-list of websites. This way I could give a student a voucher that can only access Wikipedia and maybe a few other websites. Obviously we don't want to restrict the usage of the internet in this way for everyone, but it is valid and important to control what certain people use. That is part of what defines a "commons" -- that we manage it collectively. If it is not managed, not coordinated, that is where we find the problems of uncontrolled access.

@germanferrero and @luandro , do you know who could implement this? And can you give some ideas about how it might be implemented, please?

**Comment by patogit** [comment]
I see this issue is mentioned in [this planning document](https://hackmd.io/2-uMnns3Qwub_drQWbrv6Q?view#COW-Herd-Controls) from last year, under the section COW Herd Controls. I also see some relevant conversation on [this forum thread](https://communitynetworks.group/t/traffic-control-on-openwrt-and-libremesh-second-round/888) about traffic control.

**Comment by tbdinesh** [comment]
Hello @patogit,

What we did was to come up with a scheme to get women and girls of the
community to be the voucher negotiators.
It started with the Pirania and TC available features and integration
issues. Which lead to this outcome which we thought is a better way in
the first place.
Yes, more parameters for Pirania is always useful.

an outline of this is here (look for TC activities):
https://pad.riseup.net/p/cop-keep
and the development of CoP deliverables
https://bit.ly/cop-2022-servelots-reports

@tania batata tdruiva and @shalini would be the people to acknowledge.

d
